OSDOCS-20042: adds OIDC authentication RHCL (#112868)
openshift/openshift-docs: modules/proc-rhcl-oidc-rbac-policy.adoc

// Module included in the following assemblies:
//
// *deployment/rhcl-oids-authentication.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-rhcl-oidc-rbac-policy_{context}"]
= Using RBAC with an OIDC AuthPolicy custom resource

[role="_abstract"]
You can configure role-based-access to use with your OpenID Connect (OIDC) authentication by adding the roles to your `AuthPolicy` custom resource (CR).

.Prerequisites

* You installed {prodname} on one or more clusters.
* You installed the {oc-first}.
* You have write access to the {ocp} namespaces you need to work with.
* You have access to external or on-premise DNS.
* You created a `Gateway` object.
* You created an `HTTPRoute` or `GRPCRoute` object.
* You created an `AuthPolicy` CR.
* You got a token from your OIDC provider.

.Procedure

. Add the following roles to your `AuthPolicy` CR by using the following example:
+
.Example AuthPolicy with RBAC definitions
[source,yaml]
----
apiVersion: authorino.kuadrant.io/v1
kind: AuthPolicy
metadata:

>   name: talker-api-protection

Reviewer: The [previous example](https://github.com/openshift/openshift-docs/pull/112868/changes#r3473475738) uses a placeholder. Here the name of the CR is fixed.

spec:
#...
  patterns:
    "member-role":
    - selector: auth.identity.realm_access.roles
      operator: incl
      value: member
    "admin-role":
    - selector: auth.identity.realm_access.roles
      operator: incl
      value: admin
  authorization:
    "rbac-resources-api":
      when:
      - cel: "http.path.matches('^/resources(/.*)?$')"
      patternMatching:
        patterns:
        - patternRef: member-role
    "rbac-delete-resource":
      when:
      - cel: "http.path.matches('^/resources/\\\\d+$') && http.method == 'DELETE'"
      patternMatching:
        patterns:
        - patternRef: admin-role
    "rbac-admin-api":
      when:
      - cel: "http.path.matches('^/admin(/.*)?$')"
      patternMatching:
        patterns:
        - patternRef: admin-role
----
+
* The `member` role is required for requests to `/resources[/*]`.
* The `admin` role is required for `DELETE` requests to `^/resources/\\\\d+$`.

> * The `"rbac-admin-api"` role is required for requests to `^/admin(/.*)?$`.

Reviewer: ```suggestion
* The `admin` role is required for requests to `^/admin(/.*)?$`.
```

// Module included in the following assemblies:
//
// *deployment/rhcl-oidc-authentication.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-rhcl-oidc-rbac-policy_{context}"]
= Using RBAC with an OIDC AuthPolicy custom resource

[role="_abstract"]
You can configure role-based-access to use with your OpenID Connect (OIDC) authentication by adding the roles to your `AuthPolicy` custom resource (CR).

.Prerequisites

* You installed {prodname} on one or more clusters.
* You installed the {oc-first}.
* You have write access to the {ocp} namespaces you need to work with.
* You created a `Gateway` object.
* You created an `HTTPRoute` or `GRPCRoute` object.
* You created an `AuthPolicy` CR.
* You got a token from your OIDC provider.

.Procedure

> . Add the following roles to your `AuthPolicy` CR by using the following example:

Reviewer 2: Similar comment about single step procedures:

```suggestion
* Add the following roles to your `AuthPolicy` CR by using the following example:
```

// Module included in the following assemblies:
//
// *deployment/rhcl-oids-authentication.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-rhcl-oidc-rbac-policy_{context}"]
= Using RBAC with an OIDC AuthPolicy custom resource

[role="_abstract"]
You can configure role-based-access to use with your OpenID Connect (OIDC) authentication by adding the roles to your `AuthPolicy` custom resource (CR).

.Prerequisites

* You installed {prodname} on one or more clusters.
* You installed the {oc-first}.
* You have write access to the {ocp} namespaces you need to work with.

> * You have access to external or on-premise DNS.

Reviewer 3: No need for DNS

* You created a `Gateway` object.
* You created an `HTTPRoute` or `GRPCRoute` object.
* You created an `AuthPolicy` CR.
* You got a token from your OIDC provider.

.Procedure

. Add the following roles to your `AuthPolicy` CR by using the following example:
+
.Example AuthPolicy with RBAC definitions
[source,yaml,subs="+quotes"]
----

> apiVersion: authorino.kuadrant.io/v1

Reviewer 3: I think apiVersion is just `kuadrant.io/v1` - it is like that in the other example in this PR

Reviewer: Nice catch!

kind: AuthPolicy
metadata:
  name: _<user_apps>_

>   namespace: _<api_oidc_rbac_auth>_

Reviewer: ```suggestion
  name: _<api_oidc_rbac_auth>_
  namespace: _<user_apps>_
```

>   namespace: _<oidc_apps_protection>_

Reviewer: Even if they all placeholders, can we try to standardise the names of the resources used across the examples / use names that better represent what the example is about?

| module | policy namespaced name | suggested |
|--------|------------------------|-----------|
| [`proc-set-oidc-auth-policy_{context}`](https://github.com/ShaunaDiaz/openshift-docs/blob/4a5941c79994538fbd9b6361c9b06d9490a28fca/modules/proc-rhcl-oidc-authpolicy.adoc?plain=1#L35-L36) | `<oidc_apps>/<api_oidc_auth>` | `<user-apps>/<api_oidc_auth>` |
| [`proc-set-oidc-auth-policy-verify_{context}`](https://github.com/ShaunaDiaz/openshift-docs/blob/4a5941c79994538fbd9b6361c9b06d9490a28fca/modules/proc-rhcl-oidc-authpolicy-verify.adoc?plain=1#L22) | `<oidc_apps>/<talker_api_oidc_auth>` | `<user-apps>/<api_oidc_auth>` |
| [`proc-rhcl-oidc-rbac-policy_{context}`](https://github.com/ShaunaDiaz/openshift-docs/blob/4a5941c79994538fbd9b6361c9b06d9490a28fca/modules/proc-rhcl-oidc-rbac-policy.adoc?plain=1#L33-L34) (this file) | `<oidc_apps_protection>/<api_protection>` | `<user-apps>/<api_oidc_rbac_auth>` |

Author: absolutely!

spec:
#...
  patterns:
    "member-role":
    - selector: auth.identity.realm_access.roles
      operator: incl
      value: member
    "admin-role":
    - selector: auth.identity.realm_access.roles
      operator: incl
      value: admin
  authorization:
    "rbac-resources-api":
      when:
      - cel: "http.path.matches('^/resources(/.*)?$')"
      patternMatching:
        patterns:
        - patternRef: member-role
    "rbac-delete-resource":
      when:
      - cel: "http.path.matches('^/resources/\\\\d+$') && http.method == 'DELETE'"
      patternMatching:
        patterns:
        - patternRef: admin-role
    "rbac-admin-api":
      when:
      - cel: "http.path.matches('^/admin(/.*)?$')"
      patternMatching:
        patterns:
        - patternRef: admin-role
----
+

> * Replace the `spec.metadata.name` parameter value with the name you want to use for your `AuthPolicy` CR.

Reviewer 4: ```suggestion
* Replace the `metadata.name` parameter value with the name you want to use for your `AuthPolicy` CR.
```

Same for metadata.namspace

// Module included in the following assemblies:
//
// *deployment/deployment.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-rhcl-oidc-rbac-policy_{context}"]
= Using RBAC with an OIDC AuthPolicy custom resource

[role="_abstract"]
You can configure role-based-access to use with your OpenID Connect (OIDC) authentication by adding the roles to your `AuthPolicy` custom resource (CR).

.Prerequisites

* You installed {prodname} on one or more clusters.
* You installed the {oc-first}.
* You have write access to the {ocp} namespaces you need to work with.
* You have access to external or on-premise DNS.
* You created a `Gateway` object.
* You created an `HTTPRoute` or `GRPCRoute` object.
* You created an `AuthPolicy` CR.

> * You configured got a token from your OIDC provider.

Reviewer: ```suggestion
* You obtained a token from your OIDC provider.
```

.Procedure

. Add the following roles to your `AuthPolicy` CR by using the following example:
+
.Example AuthPolicy with RBAC definitions
[source,yaml]
----
apiVersion: authorino.kuadrant.io/v1

> kind: AuthPolicy

Reviewer: Let's simplify this sample AuthConfig using the short version Well Known Attributes (e.g. `request.path`) and CEL predicates instead of selector/operator/value tuples in the `when` conditions.

  name: talker-api-protection
spec:
#...
  patterns:
    "member-role":
    - selector: auth.identity.realm_access.roles
      operator: incl
      value: member
    "admin-role":
    - selector: auth.identity.realm_access.roles
      operator: incl
      value: admin
  authorization:
    "rbac-resources-api":
      when:
      - selector: context.request.http.path
        operator: matches
        value: ^/resources(/.*)?$
      patternMatching:
        patterns:
        - patternRef: member-role
    "rbac-delete-resource":
      when:
      - selector: context.request.http.path
        operator: matches
        value: ^/resources/\d+$
      - selector: context.request.http.method
        operator: eq
        value: DELETE
      patternMatching:
        patterns:
        - patternRef: admin-role
    "rbac-admin-api":
      when:
      - selector: context.request.http.path
        operator: matches
        value: ^/admin(/.*)?$
      patternMatching:
        patterns:
        - patternRef: admin-role
----
+
* The "rbac-resources-api" role is required for requests to `/resources[/*]`.
* The `"rbac-delete-resource"` role is required for `DELETE` requests to `/resources/{id}.`

> * The `"rbac-admin-api"` role is required for requests to `/admin[/*]`.

Reviewer: ```suggestion
* The "member" role is required for requests to `/resources[/*]`.
* The `"admin"` role is required for `DELETE` requests to `/resources/{id}.`
* The `"admin"` role is required for requests to `/admin[/*]`.
```
